src/CMakeLists.txt

```diff
@@ -9,6 +10,18 @@ file(GLOB INL ../include/*.inl *.inl)
 file(GLOB_RECURSE EXTERNAL_INC ../external/*.h ../external/*.hpp)
 
 add_library(d3d11on12 SHARED ${SRC} ${INC} ${INL} ${EXTERNAL_INC} d3d11on12.rc d3d11on12.def)
+
+#if we don't have access to the d3d12translationlayer_wdk project, pull it from github (it's a part of the d3d12translationlayer repo)
+if(NOT TARGET d3d12translationlayer_wdk)
+    message("Fetching D3D12TranslationLayer from GitHub...")
+    FetchContent_Declare(
+        d3d12translationlayer
+        GIT_REPOSITORY https://github.com/microsoft/D3D12TranslationLayer.git
+        GIT_TAG        origin/master

```

Comment:
I'm wondering if we should go ahead and freeze this to the current `HEAD`, so changes to the `D3D12TranslationLayer` project can't accidentally break this? I'm torn - it does mean that if there is a break, and then we want to pick up a new version, we'll have to address that break at that time.

Reply:
I went back and forth on that as well... maybe a solution would be to have a branch for the last known good commit? Ideally that would be updated by a git workflow or something. At the same time, this currently only runs the first time someone builds the project without the translation layer. This means that if we break something after they've pulled it's not an issue. It also means that if they do pull a broken state of the repo it's more annoying to update, and even if the pull a good version it will eventually go stale... I've been thinking about that off and on since I pushed the change yesterday, and I'm not sure what the best path forward is. I see that you have a comment related to this, so I'll reply with some thoughts there.

Reply:
> At the same time, this currently only runs the first time someone builds the project without the translation layer.

What I saw while reading the docs was that this will happen every time the project is reconfigured. Which... I guess still doesn't happen that often - unless it's actually talking about a regenerate which would happen any time you change a build file.

I think I'm inclined to pick a SHA here unless we have a good reason not to.

Reply:
What I meant by the quoted section is that, since we have `if(NOT TARGET d3d12translationlayer_wdk)` right now, once the project is pulled from git hub that target will be defined so it won't try to pull again.

I think I agree with you about the SHA though. Just read over some of the ExternalProject_Add documentation and it looks like using a specific SHA is the recommended approach:

> In general, however, specifying a commit hash should be preferred for a number of reasons:
>* If the local clone already has the commit corresponding to the hash, no git fetch needs to be performed to check for changes each time CMake is re-run. This can result in a significant speed up if many external projects are being used.
>* Using a specific git hash ensures that the main project's own history is fully traceable to a specific point in the external project's evolution. If a branch or tag name is used instead, then checking out a specific commit of the main project doesn't necessarily pin the whole build to a specific point in the life of the external project. The lack of such deterministic behavior makes the main project lose traceability and repeatability.

Reply:
> What I meant by the quoted section is that, since we have if(NOT TARGET d3d12translationlayer_wdk) right now, once the project is pulled from git hub that target will be defined so it won't try to pull again.

No, since this logic is run every time the project is reconfigured. When you run the configure step a second time, by the time the script reaches this line, there is not yet a target with that name, because `add_subdirectory` hasn't been invoked, so the `MakeAvailable` call will run again - which will check to see if what's downloaded is up-to-date and then add the project.

Reply:
Aaaah you're right, and that makes sense too. Not sure how I got that misconception, thanks for clearing it up though.